### ISSUE TITLE
New `windows_event_log` source

### ISSUE THREAD

**Issue by ghost** [vectordotdev/vector#1206]
It might be useful to support [Event Log](https://docs.microsoft.com/en-us/windows/win32/wes/windows-event-log) as a source on Windows platform.

As a first step, we need a specification to better understand requirements.

**Comment by mhf-ir** [comment]
It's could be awesome, we use in our data centers for collection windows machines.

**Comment by jamtur01** [comment]
We should also consider a stand-alone transform for the Windows Event Log format and evtx files.

**Comment by ethack** [comment]
I'd really like to see Vector as an alternative to Winlogbeat for Windows log shipping.

This library could be useful (not sure about license compatibility):
https://github.com/omerbenamram/evtx

**Comment by ypid-geberit** [comment]
Also refer to the discussion in the duplicate issue #2719. To make it short, I see adding support for the Windows event log as a can of worms. Winlogbeat already does a good job.

**Comment by ethack** [comment]
@ypid-geberit I read [your comment](https://github.com/vectordotdev/vector/issues/2719#issuecomment-881475825) and looked at the beats issue you referenced. I'd like to offer an alternate opinion. 

Based on my understanding of the beats issue, those GUID mappings are desirable but not actually necessary in order to have basic Windows event log ingestion. [This comment](https://github.com/elastic/beats/issues/16301#issuecomment-585751674) and the following one indicate that simply ingesting Windows events was possible prior to the issue being opened and the motivation was convenience.

The only [linked pull request](https://github.com/elastic/beats/pull/20684/files#diff-4ec1cb8494d34109141faddb82b4c0145f2774b9a5aa9c1f499fe75a4e34be08) I could find shows what sort of changes would go into the GUID mappings. Note that the code in Winlogbeats is in JavaScript rather than Go. Likewise, I could see something like the mappings being implemented in VRL remaps or a Lua transform. By putting the base XML/EVTX parsing in place in Rust, you'd open up the ability for users to start implementing these types of convenience mappings and potentially contributing these to the community. I agree with you this is a can of worms, but I also don't see it as a requirement in order for Vector to have something usable.

> GUID lookups and so on that are best done by the agent on the system that emits the logs.

I don't see where it's a requirement to do these on the log emitter. The mappings seem to be hard-coded and not something that requires environment specific queries. I could be wrong and I'm not even sure if that is what you meant by "best". That being said it would actually be pretty cool to be able to do live queries against Active Directory for enrichment in some log sources where a SID is available but not a user/group name. But that's a separate issue.

As for the benefits of re-creating the wheel when Winlogbeat already exists, I don't have much more to add than [clay584's comment](https://github.com/vectordotdev/vector/issues/2719#issuecomment-772768256). Just wanted to link that to the discussion here since the other duplicate issue is closed.

**Comment by ypid-geberit** [comment]
All right. I see your point and you also got my point. A Windows event log source can be done with Vector. But some form of enrichment will be needed with the event log, at least for Windows internal logs. This is different to other logs so I wanted to point this out beforehand. AD lookup could be interesting.

But due to a change in my work responsibilities (no Microsoft anymore 😉), I guess I will drop out of this discussion.

**Comment by Rushan4eg** [comment]
How all the going with the progress here?

**Comment by jszwedko** [comment]
> just trying to get around the issue, would it be possible to pull the xml files for the windows event logs instead, and use that as a work around?

Hey @umpa385 . That could work. There is a `parse_xml` function in VRL: https://vector.dev/docs/reference/vrl/functions/#parse_xml

**Comment by umpa385** [comment]
will report back on how to get that done as a potential work around even though a native solution would be best. 

considering using this https://github.com/omerbenamram/evtx#example-usage-as-library

**Comment by umpa385** [comment]
Hello wanted to add a current work around that I tested (its not a fully supported solution, but works for my use case) 

Using https://github.com/omerbenamram/evtx as a way to convert the evxt files into json and then having vector pick those json logs up. Please feel free to comment or reach out here if anyone want more details on this.

**Comment by tshepang** [comment]
@umpa385 do you maintain a fork of Vector to do this

**Comment by umpa385** [comment]
> @umpa385 do you maintain a fork of Vector to do this

Nope I have a local repo that I'm using to convert the evtx files to json (kind of hacky but thinking about running the .exe I mentioned in the above comment via task scheduler to convert the data to json and then have vector pick that up). That is the way to do it "locally". 

The actual way I plan to implement this possibly, is seeing if I can just push the raw evtx files to a vector aggregator and then do the conversion from evtx to json on the aggregator, so that way less overhead on the endpoint itself. 

Let me know if that makes sense or if you have a better way to do this. More than willing to give feedback etc. to help build out something native for windows.

**Comment by prognant** [comment]
To listen for windows event in realtime we could probably use https://docs.rs/winapi/latest/winapi/um/winevt/fn.EvtSubscribe.html, I'm not exactly sure what we would get from that, but the evtx lib if needed or just plain xml parsing we should probably be able to get a native source working.

**Comment by umpa385** [comment]
so wanted to add more to this. I was able to get this to kind of work in a hacky way, that isn't always 100%.

You use vector to push to the raw evtx file to a aggregator and then run the https://github.com/omerbenamram/evtx#example-usage-as-library binary to convert the evtx files to json and then reginest where you want them to go. 

The issues being using vector to pull the evtx files from windows sometimes causes corruption on the evtx files itself 
This leads to the rust binary for the convert to not work correctly leading to blank conversions. 

I was able to get this working correctly for a test env, but when trying to go into prod it fell apart, because of the issue mentioned above.

**Comment by backeby** [comment]
This would be an amazing feature to have

**Comment by wikro** [comment]
Yes, this is something we would want too!

**Comment by ebusto** [comment]
This is relevant to my interests and I wish to subscribe to your newsletter.

**Comment by otisg** [comment]
@jszwedko is this still on the roadmap?  If so, what sort of timeline are you thinking?  Thanks.

**Comment by jszwedko** [comment]
> @jszwedko is this still on the roadmap? If so, what sort of timeline are you thinking? Thanks. @umpa385 thank you for sharing your approach and diligently following up with your updates. 👍

Unfortunately no updates at this point. We'd be happy to see a PR in this area if there is a motivated contributor. We are partially hampered by the lack of Windows expertise on the team.

**Comment by umpa385** [comment]
I took another stab at this, I have a semi working version if you don't care about powershell logs. The idea is to use 
https://learn.microsoft.com/en-us/windows-server/administration/windows-commands/wevtutil to export the logs and then use the file sink to pick them up via vector and send it where you want it. 

Its not the best process, but create a power shell script to put the files in directory and then have vector pickup those files and have powershell overwrite every time (I was going to use a task schedule to write out the logs every 3 mins). 

I'm going to try this again with winlogbeat to file instead so I can also get the powershell logs.

**Comment by otisg** [comment]
@umpa385 is wevtutil available in "all" versions of Windows?

**Comment by umpa385** [comment]
Okay so sadly takes 2 agents, but it works, use winlogbeat to send logs over to vector. Trying to do something native is much harder right now, but this two agent system works.

**Comment by umpa385** [comment]
Here is the config I have in vector. Basically setup winlogbeat as a logstash output and then vector ingests that. 

`sinks:
  win_logs:
    type: http
    inputs:
      - windows_logs
    encoding:
      codec: "json"
    uri: >-
      https:
    auth:
      strategy: bearer
      token: 
    compression: gzip
    batch:
      max_bytes: 1000000
  sink_id:
    type: "aws_s3"
    inputs:
      - windows_logs
    bucket: 
    key_prefix: 
    auth:
      region: 
      access_key_id: 
      secret_access_key:
      assume_role: 
    encoding:
      codec: "json"`

**Comment by umpa385** [comment]
sinks:
  win_logs:
    type: http
    inputs:
      - windows_logs
    encoding:
      codec: "json"
    uri: >-
      https:
    auth:
      strategy: bearer
      token: 
    compression: gzip
    batch:
      max_bytes: 1000000
  sink_id:
    type: "aws_s3"
    inputs:
      - windows_logs
    bucket: 
    key_prefix: 
    auth:
      region: 
      access_key_id: 
      secret_access_key:
      assume_role: 
    encoding:
      codec: "json"

**Comment by mkfrey** [comment]
> Okay so sadly takes 2 agents, but it works, use winlogbeat to send logs over to vector. Trying to do something native is much harder right now, but this two agent system works.

Nice idea. I've tried it and it works for basic transfer of windows logs. However, I've found that with the default mem queue, messages are delayed for up to 40 seconds. When using a disk queue for robustness, this can even go up to 10 minutes.

So having a log source directly integrated would be a huge improvement.

**Comment by umpa385** [comment]
> > Okay so sadly takes 2 agents, but it works, use winlogbeat to send logs over to vector. Trying to do something native is much harder right now, but this two agent system works.
> 
> Nice idea. I've tried it and it works for basic transfer of windows logs. However, I've found that with the default mem queue, messages are delayed for up to 40 seconds. When using a disk queue for robustness, this can even go up to 10 minutes.
> 
> So having a log source directly integrated would be a huge improvement.

Have you tried to use the logstash source? We are not seeing that level of delay, we are not writing it down to a file.

**Comment by mkfrey** [comment]
Yes, I used the logstash output of winlogbeat and the logstash source with Vector. Winlogbeat without disk introduces a delay ranging from 0 to 30 seconds.

**Comment by sandervandegeijn** [comment]
Since Elastic is creating somewhat of a walled garden since the licensing change I'd rather stay away from winlogbeat and use Vector natively. Alternative could be to use fluent-bit, but I've had issues with it in the past, so I'm hesitant. A native Vector solution would be nice.

**Comment by 2blackpanther** [comment]
@jszwedko 
@pabloem 
Any updates on a native vector solution?

**Comment by mkfrey** [comment]
I am currently trying to get some stuff done. However, it's not very simple:

- The windows APIs are very annoying to use and it's difficult to get any information on best practices (e.g. whether using push or pull is better).
- A checkpointer for bookmarks needs to be implemented.
- Documentation for implementing a plugin is a bit lacking in my opinion. I have to make my way along by looking at existing plugins and some docs.

**Comment by sandervandegeijn** [comment]
Couldn't you get some inspiration from fluentbit? They have a winevent source :)

**Comment by mkfrey** [comment]
I did, and I even found a minor mistake there. 😁

**Comment by nikriaz** [comment]
Winlogbeat works actually nice. Up to 30 sec delay is expected, since Logstash sampling interval output default is 30s (configurable). 

This is working example for IIS logs:

 winlogbeat.yml
```
winlogbeat.event_logs:
  - name: Microsoft-IIS-Logging/Logs
    ignore_older: 72h
    tags: [iis_etw]

output.logstash:
  enabled: true
  hosts: ["HOST.domain.local"]
  escape_html: true

logging:
  level: warn
  to_files: true
  files:
    path: D:\logs\Winlogbeat
    name: winlogbeat.log
    keepfiles: 7
```

vector.yaml (with emitting cleaned actual Windows event content, all other noise deleted)
```
sources:
  windows_logs:
    type: logstash
    address: 0.0.0.0:5044

  iis_clean:
   type: remap
   inputs:
     - windows_logs
   source: |-
      ed = .winlog.event_data
      clean = {}
      clean.host = .host.name
      clean.date = ed."date"
	  ... all other fields...
      . = clean
	  
sinks:
  debug_sink:
    type: console
    inputs: 
     - iis_clean
    target: stdout
    encoding:
      codec: json
```

**Comment by umpa385** [comment]
I've been running the config I have above in production for over a year it works as long as you are okay with the two agent method. 

Something native for windows would be ideal but this isn't bad either and works for all the windows logs we need. The added benefit of using vector is we split out our logs to a cheaper log source for any longer retention periods to an s3 bucket which is really nice.

**Comment by mkfrey** [comment]
Unfortunately I haven't had much time in the past months to continue working on a plugin implementation. 

I have uploaded the latest version of my windows event log code here:  https://github.com/mkfrey/windows-event-rs-testing

So if anyone wants to go on from there, feel free to try it. Maybe I will be able find some time in the future.

**Comment by pront** [comment]
> Any updates on this from the core team? We are considering implementing, and this is a pretty serious blocker for us.

Hello, this is not prioritized internally. We would be delighted to review a PR though.

**Comment by tot19** [comment]
Roger that. Thank you!

**Comment by HarishHary** [comment]
Any traction on this?

**Comment by inferno-umar** [comment]
@tot19 are u implementing it ?

**Comment by tot19** [comment]
Hello @inferno-umar. I am not implementing this.

**Comment by r3code** [comment]
As a workaround instead of vector you can use fluentbit for collecting Windows event log (https://docs.fluentbit.io/manual/data-pipeline/inputs/windows-event-log). Vector can receive fluentbit logs https://vector.dev/docs/reference/configuration/sources/fluent/

**Comment by HarishHary** [comment]
@r3code You would pretty much install 2 agent on the endpoint to collect logs right? @tot19 no traction from your team at the end? Too bad we don't have it? Anyone knows where to begin with if we need to implement it?

I wonder how we keep up with the log format changes, log version with all the flavors of windows. It does have a lot of impacts on logs.

**Comment by mkfrey** [comment]
What do you mean by log format changes? Windows version has basically no impact on logs. The API is the same since Vista.

**Comment by sandervandegeijn** [comment]
If you format the raw logs, yep. Even winlogbeat does that?

This would help us to standardize on Vector across the infrastructure. For Linux/containers vector is almost a nobrainer, I have very mixed experiences with FluentBit, encountered serious bugs that have not been fixed for years. Vector on the other hand is fantastic so far.

But unusable on Windows because we really really need the event logs...

**Comment by sandervandegeijn** [comment]
Okay, I'm feeling like Alice in wonderland here. I have a vector on my Windows machine that ingests Windows event logs..

To set expectations, I am no Rust developer at all: I'm reasonable at Python, I can get most things done in C# and dabble in other languages. So developing this myself was in no way possible in the short term. Rust is quite complex and I have no immediate use for it.

But...I did however wanted to see how good Anthropics Claude Code is. So....I set it to work. It cost ±24 hours of the LLM chewing on the code and my workstation on compiling, but I have a version of vector here that is ingesting Windows events. The fact that it works, compiles and has running unit tests amazed me.  Output of the logs looks ok at first glance, but I want to take a good look at it and compare it to i.e. Fluentbit. 

I tried to guide it with the development guidelines for the Vector project, let it analyse and compare other sources within the Vector project to evaluate the code quality of it's own creation. Can someone that can code in Rust take a look at the code to see if it's usable *at all*? Maybe it can serve as a basis for the real implementation.

https://github.com/sandervandegeijn/vector

If not, at least it was a fun experiment. :)

**Comment by mkfrey** [comment]
Some parts look okay, some are hot mess. E.g. there is a lot of code around storing and loading bookmarks, but they are actually never used.

The first events are fetched using `EvtQuery`, which does not even support using bookmarks. `EvtSubscribe` would be more fitting, it also allows to wait on a signal for new messages instead of the need to poll.

Also it heavily relies on rendering the message as XML and then parsing that, instead of rendering most of the stuff as raw EvtVariants. Entirely valid, but not as flexible.

But there are still some parts in there which look usable.

**Comment by r3code** [comment]
@HarishHary Another workaround. 
Use https://learn.microsoft.com/en-us/windows-server/administration/windows-commands/vevtutil to export periodically an eventlogs to file , use vector to watch this file and parse with transforms.

**Comment by HarishHary** [comment]
@mkfrey  > What do you mean by log format changes? Windows version has basically no impact on logs. The API is the same since Vista.

API might have stayed the same but I know some logs that changed/evolved depending on versions and so on? Either they added more info or removed some. In any case, I just wonder if we need to test on all the flavors of Windows to have something stable? I have no idea. Just thinking out loud here, I might be wrong.

@sandervandegeijn Wonderful first work here... Unfortunately, I have 0 skill in rust... 

@r3code Thx for the workaround... But installing and maintaining 2 agents is not something I am willing to do for now. Here is more like a schedule task if I understood you correctly. This will also increase collection delay right? Thanks again for the suggestion.

Regards,

**Comment by sandervandegeijn** [comment]
Based on the feedback (thanks @mkfrey ) I did another round with Claude. I've been quite impressed with Claude Code for my own projects, but in those cases I was able to provide better guidance to Claude and steer it in the right direction more. But as for the new run: it compiles, the tests complete, it's now based on the event subscription api and it should contain less bullshit code. Output looks quite good.

I'm a bit out of my depth here, so if this is a futile exercise: no problem. That's the nature of these LLM's, it's hit or miss. If it's a basis someone more capable than me can work of: great :) I'll leave the repo online and will iterate further if requested but with the notion that I'm just trying and doing an experiment. Just open an issue on the forked repo and we'll see.

**Comment by pront** [comment]
If anyone is interested in the future: https://github.com/vectordotdev/vector/compare/master...sandervandegeijn:vector:master (thanks @sandervandegeijn)

**Comment by ghost** [comment]
Hi @prognant and @sandervandegeijn, is the pull request you linked previously still in development? If not, would a pull-request that implements a Windows Event Log source still be considered?

**Comment by tot19** [comment]
@wssheldon I am not the original poster, but it seems like everyone here is open to a PR.

This would make my life 10x easier, as it will allow me to deploy one single logging agent on all of my Windows boxes. Having to deploy different agents for app and system logs currently.

@pront @sandervandegeijn What is still outstanding from the branch above?

**Comment by tot19** [comment]
I don't have Windows machines to test with, but trying to help where I can. I made clippy happy and rebased.

[updated branch](https://github.com/tot19/vector/tree/windows)

**Comment by sandervandegeijn** [comment]
Thanks Tot, what did you think of the code after your fixes? Would this be usable and merit further testing?

I do have Windows machines to my disposal at the university.

**Comment by tot19** [comment]
I think this absolutely merits further testing. Tracked down a windows system and can confirm the event log source is working as expected.

Only issue I see now is pretty significant load when a user signs on (via OpenSSH in my case). I am running a single CPU 2022 Core VM without any other active apps running.

**Comment by tot19** [comment]
FYI added a few items on my branch:
- Fixed a couple of bugs
- Added rate limiting
- Added the ability to resume from a checkpoint.

Curious if anyone here is more Rust or Windows savvy and want to try putting it through its paces? Not sure what else is needed before it's ready for a PR.

**Comment by tot19** [comment]
Opened a PR #24305 ... Please note, I _am not_ a Rust developer or Windows person. I had Claude code write this as I performed tests against my Windows systems. Would appreciate feedback and testing.

**Comment by tot19** [comment]
Hello, checking to see if anyone has had a chance to give this a spin or check out the code?

**Comment by mithr4ndir** [comment]
I would love to see this supported on windows clients. I can do some testing if needed. Currently the MacOS and Linux endpoints use vector. I want to centralize on just one log shipper. 

What do we need to do to get eyes on this from Vector devs?

**Comment by sandervandegeijn** [comment]
I don't have any connections to the developers, can we contact them directly? We could open a pull request, but it'll probably be declined and I don't know if we have the expertise in this group to fix the issues?

**Comment by tot19** [comment]
Hey, thank you both for the interest. I did open a PR for this and reached out to the developers on Discord.

Still waiting to hear back. Thanks!

**Comment by nf-brentsaner** [comment]
@sandervandegeijn @tot19 FWIW, the responses might be slow but this is one of the repos where the maintainers/owners are *really* interested in getting your PR accepted. i don't think i've seen a single PR rejected. as long as you're willing to work with them it should eventually be accepted in.

**Comment by sandervandegeijn** [comment]
Hoping for some progress on this :) We are fully committing to using Vector on our kubernetes and Linux ecosystem. I'd really like to standardize the Windows ecosystem on Vector as well :)

**Comment by tot19** [comment]
Hey @pront, possible you could take a look at this when you get a chance? Thanks!

**Comment by pront** [comment]
> Hey [@pront](https://github.com/pront), possible you could take a look at this when you get a chance? Thanks!

Hi @tot19, I will take a look since this is a highly request features. However, it's challenging for me to do meaningful testing on real windows machines. Perhaps we can add an integration test, see how we run windows tests here: [.github/workflows/unit_windows.yml](https://github.com/vectordotdev/vector/blob/master/.github/workflows/unit_windows.yml)

**Comment by tot19** [comment]
Thanks for the review @pront ! I’ve addressed the feedback and made a few related improvements:

- Removed internal #[cfg(windows)] / #[cfg(not(windows))] gates under src/sources/windows_event_log/* so the module‑level gate in mod.rs is the single source of truth.
- Added windows-event-log-integration-tests to all-integration-tests in Cargo.toml.
- Added integration config at test.yaml.
- Added a Windows CI hook to run the integration test via make test-integration-windows-event-log in unit_windows.yml.
- The integration test seeds a log entry via eventcreate to ensure data is present (Windows‑only, Admin required).

On the docker‑compose point: I’m not sure how to provide a containerized environment for Windows Event Log. Would you prefer a compose.yaml stub anyway, or is a Windows‑hosted integration run acceptable here? If there’s a pattern you want me to follow, I’m happy to align.

I’m also trying to get access to a Windows machine so I can run the integration suite and report results.

https://github.com/vectordotdev/vector/pull/24305/changes/4e5de22b391d7175e77f970f57f6985f1dd8b8bf..eb87b8fa70d301a3605967965d09511500223e87

**Comment by pront** [comment]
> Thanks for the review [@pront](https://github.com/pront) ! I’ve addressed the feedback and made a few related improvements:
> 

Thank you for addressing the review feedback so quickly. I want to get my hands on a Windows dev environment so I can review Windows PRs more effectively
 
> On the docker‑compose point: I’m not sure how to provide a containerized environment for Windows Event Log. Would you prefer a compose.yaml stub anyway, or is a Windows‑hosted integration run acceptable here? If there’s a pattern you want me to follow, I’m happy to align.
> 
> I’m also trying to get access to a Windows machine so I can run the integration suite and report results.
> 
> [#24305 (changes)](https://github.com/vectordotdev/vector/pull/24305/changes/4e5de22b391d7175e77f970f57f6985f1dd8b8bf..eb87b8fa70d301a3605967965d09511500223e87)

Regarding testing, we don't have to follow the pattern from other integration tests. We can run Windows test on a GitHub hosted runner similar to how we run the unit tests for Windows: https://github.com/vectordotdev/vector/blob/master/.github/workflows/unit_windows.yml#L17. Creating an `integration_windows.yml` might be a good idea:

```yaml
jobs:
    test-integration-windows:
      runs-on: windows-2025 # https://github.com/actions/runner-images/tree/main/images/windows

      steps:
        # checkout, bootstrap, build

        - name: Run integration test
          run: make test-integration-${{ inputs.test_name }}

        # ...
```

**Comment by tot19** [comment]
Thanks @pront ! I added a dedicated Windows integration workflow that runs the Windows Event Log integration test on a GitHub Windows runner (and removed it from the unit job). Let me know how that looks!

Integration test with detailed output on most available windows image: https://github.com/tot19/vector/actions/runs/22498415565/job/65178884878

Integration test pending windows 2025 with 8 cores and no debug logs: https://github.com/tot19/vector/actions/runs/22498501267/job/65179179573

**Comment by tot19** [comment]
Ok, glad you suggested the integration tests. Found some major issues with logs being dropped under high contention. Added ~30 integration tests and restructured some major components of the code to resolve that issue (biggest being pulling logs via event subscription rather than using callback based push model). I am officially done touching things. Thanks!

**Comment by pront** [comment]
Thank you @tot19 for your work on this highly requested feature!

Note to the community: 
* This source is a beta feature. 
* Please create new issues if you find bugs. Also, please include as many details as possible especially about the environment.

Closing per: https://github.com/vectordotdev/vector/pull/24305#event-23557174747

**Comment by sandervandegeijn** [comment]
Fantastic!! Just curious, was my first experiment helpful and did any of it make it into the merge?  Although I'm decent in other languages, I'm not familiar with rust, so if it was useful it does prove some of the added value of these agentic coding agents.

**Comment by tot19** [comment]
Hey @sandervandegeijn, it was very helpful in getting me started and giving me some confidence to take on the effort. The code itself was functional; however, I had to change approaches to how the event logs were consumed due to performance limitations (pull vs poll) ... Because of that, the initial code was largely not utilized. That being said, I still greatly appreciate you getting this kicked off and the help!

**Comment by sandervandegeijn** [comment]
Cool thanks :) Glad it helped somewhat :) Now let's see how it will be used in the field.
